File: PrefixSum/E724/E724.py
class Solution:
    def pivotIndex(self, nums: List[int]) -> int:
        total = 0
        for num in nums:
            total += num

        # Create prefix from the left and compare with the right prefix(total)
        leftPrefix = 0
        for i in range(len(nums)):
            if (total - nums[i] - leftPrefix) == leftPrefix:
                return i
            leftPrefix += nums[i]
        return -1

#! 3min 45sec (After doing the question in ts)

File: PrefixSum/E724/E724.ts
/*
Find Pivot Index

Given an array of integers nums, calculate the pivot index of this array.

The pivot index is the index where the sum of all the numbers strictly to the left of the index is equal to the sum of all the numbers strictly to the index's right.

If the index is on the left edge of the array, then the left sum is 0 because there are no elements to the left. This also applies to the right edge of the array.

Return the leftmost pivot index. If no such index exists, return -1.
*/

function pivotIndex(nums: number[]): number {
	// Find the prefix sum where you minus the value at that index, the max sum is twice the value of the prefix sum running up to that index
	const prefix: number[] = [nums[0]];
	for (let i = 1; i < nums.length; i++) {
		prefix.push(nums[i] + prefix[i - 1]);
	}
	const maxSum: number = prefix[prefix.length - 1];

	for (let i = 0; i < nums.length; i++) {
		const left: number = prefix[i] - nums[i];
		const right: number = maxSum - prefix[i];
		if (left === right) return i;
	}
	return -1;
}

//! 12min 45sec
